simpa_tests/automatic_tests/test_bandpass_filter.py

```diff
@@ -15,55 +14,152 @@
-                                             combined_time_series, filtered_time_series]):
+            for i, time_series in enumerate([self.base_time_series, self.low_freq_time_series, self.high_freq_time_series, self.combined_time_series, filtered_time_series]):
                 plt.subplot(5, 1, i + 1)
                 plt.title(labels[i])
-                plt.plot(t_values, time_series)
+                plt.plot(self.t_values, time_series)
             plt.tight_layout()
             plt.show()
+
+    def visualize_filtered_spectrum(self, filtered_spectrum: np.ndarray):
+        if Visualize:
+            plt.plot(self.frequencies, filtered_spectrum, label="filtered")
+            plt.xlabel("frequency")
+            plt.legend()
+            plt.show()
+
+    def test_tukey_filter_with_random_signal(self):
+
+        filtered_signal = tukey_bandpass_filtering(self.noisy_signal, self.time_spacing, self.cutoff_lowpass, self.cutoff_highpass, tukey_alpha=0.5)
+
+        # compute frequency spectrum
+        FILTERED_SIGNAL = np.fft.fftshift(np.fft.fft(filtered_signal))
+
+        # expected to be close to zero outside of the band
+        high_cutoff_point = self.cutoff_highpass*self.time_spacing
+        low_cutoff_point = self.cutoff_lowpass*self.time_spacing
+        assert np.allclose(0, FILTERED_SIGNAL[self.frequencies < -low_cutoff_point], atol=1e-5)
+        assert np.allclose(0, FILTERED_SIGNAL[self.frequencies > low_cutoff_point], atol=1e-5)
+        assert np.allclose(0, FILTERED_SIGNAL[np.where(np.logical_and(self.frequencies>-high_cutoff_point, self.frequencies<high_cutoff_point))], atol=1e-5)
+    
+        # expected to be not zero within the band
+        assert not np.allclose(0, FILTERED_SIGNAL[np.where(np.logical_and(self.frequencies>-low_cutoff_point, self.frequencies<-high_cutoff_point))], atol=1e-5)
+        assert not np.allclose(0, FILTERED_SIGNAL[np.where(np.logical_and(self.frequencies>high_cutoff_point, self.frequencies<low_cutoff_point))], atol=1e-5)
+
+        self.visualize_filtered_spectrum(FILTERED_SIGNAL)
+
+    def test_butter_filter_with_random_signal(self):
+
+        filtered_signal = butter_bandpass_filtering(self.noisy_signal, self.time_spacing, self.cutoff_lowpass, self.cutoff_highpass, order=9)
+
+        # compute frequency spectrum
+        FILTERED_SIGNAL = np.fft.fftshift(np.fft.fft(filtered_signal))
+
+        # expected to be close to zero outside of the band with some tolerance margin
+        high_cutoff_point = self.cutoff_highpass*self.time_spacing
+        low_cutoff_point = self.cutoff_lowpass*self.time_spacing
+        
+        assert np.allclose(0, FILTERED_SIGNAL[self.frequencies < -low_cutoff_point*1.1], atol=1)
+        assert np.allclose(0, FILTERED_SIGNAL[self.frequencies > low_cutoff_point*1.1], atol=1)
+        assert np.allclose(0, FILTERED_SIGNAL[np.where(np.logical_and(self.frequencies>-high_cutoff_point/2, self.frequencies<high_cutoff_point/2))], atol=1)
```

Comment:
These 3 asserts seem to not be reliably True for different runs (different RNG seeds). This is probably due to the random noisy signal. Maybe you can take a look at this.

Reply:
That's true. Depending on the random noisy signal, the filter sometimes works better or worse, so finding a suitable tolerance level is difficult as the worst cases are so bad. Of course we can set the tolerance level very high, but then we don't need a test anymore. As a solution I've seeded the random generator - however, there was a mistake in the code leading to non-deterministic behaviour. Now it should be fixed and the test should always pass.